.github/workflows/publish_charm.yaml

```diff
@@ -131,11 +131,54 @@ jobs:
             IMAGES=$(ls *.Dockerfile 2> /dev/null | sed s/\.Dockerfile// |  jq -Rsc '. / "\n" - [""]')
           fi
           echo "IMAGES=$IMAGES" >> $GITHUB_ENV
+  get-rocks:
+    name: Get rocks
+    runs-on: ubuntu-22.04
+    outputs:
+      rock-paths: ${{ steps.gen-rock-paths-and-images.outputs.rock-paths }}
+      images: "${{ steps.gen-rock-paths-and-images.outputs.images }}"
+    steps:
+      - name: Validate inputs
+        run: |
+          if [ "${{ inputs.cache-action }}" != "save" ] && [ "${{ inputs.cache-action }}" != "restore" ]; then
+            echo "Invalid value for cache-action. It must be 'save' or 'restore'"
+            exit 1
+          fi
+      - uses: actions/checkout@v4.1.1
+      - name: Generate rock paths and images
+        id: gen-rock-paths-and-images
+        uses: actions/github-script@v6.4.1
+        with:
+          script: |
+            const path = require('path')
+            const workingDir = core.getInput('working-directory')
+            const rockcraftGlobber = await glob.create(
+              path.join(workingDir, '**/rockcraft.yaml')
+            )
+            const rockPaths = []
+            const images = []
+            for (const rockcraftFile of await rockcraftGlobber.glob()) {
+              const rockPath = path.relative('.', path.dirname(rockcraftFile))
+              rockPaths.push(rockPath)
+              core.info(`found rockcraft.yaml in ${rockPath}`)
+              const fileHash = await glob.hashFiles(path.join(rockPath, '**'))
+              const rockName = (
+                await exec.getExecOutput('yq', ['.name', rockcraftFile])
+              ).stdout.trim()
```

Comment:
Should we do this in `github-script` instead of using an exec statement to help with readability?

Reply:
Since it was needed to get the exact same thing that was done for building rocks, I just copied the script from [here](https://github.com/canonical/operator-workflows/blob/efb823fe069df4a4e8e5a1b9f891cc17931ce960/.github/workflows/build_rocks.yaml#L69).

I believe this could be done in a follow-up PR as a refactoring. Just to make sure that these changes here are enough to fix the problem.